react/AppIcon/test/AppIcon.spec.js

```diff
@@ -0,0 +1,45 @@
+'use strict'
+/* eslint-env jest */
+
+import React from 'react'
+import { shallow } from 'enzyme'
+
+import AppIcon from '../'
+
+describe('AppIcon component', () => {
+  const app = {
+    links: {
+      icon: 'apps/test/icon'
+    }
+  }
+
+  const successFetchIcon = jest.fn().mockResolvedValue('data://test-url')
+
+  it(`renders default when fetch error occurs`, () => {
+    const failureFetchIcon = jest.fn().mockImplementation(() => {
+      throw new Error('Mocked error')
+    })
+    const component = shallow(
+      <AppIcon app={app} fetchIcon={failureFetchIcon} />
+    ).getElement()
+    expect(component).toMatchSnapshot()
+  })
+
+  it(`renders as loading`, () => {
+    const component = shallow(
+      <AppIcon app={app} fetchIcon={successFetchIcon} />
+    ).getElement()
+    expect(component).toMatchSnapshot()
+  })
+
+  it('renders app icon after fetching', done => {
+    const wrapper = shallow(<AppIcon app={app} fetchIcon={successFetchIcon} />)
```

Comment:
Minor: I think getting rid of the setTimeout would be better.

http://lea.verou.me/2016/12/resolve-promises-externally-with-this-one-weird-trick/

```js
function deferred() {
	let res, rej

	const promise = new Promise((resolve, reject) => {
	    res = resolve;
		rej = reject;
	});

	promise.resolve = res
    promise.reject = rej
	return promise;
}

it('renders app icon after fetching', done => {
	const def = deferred()
	const wrapper = shallow(<AppIcon app={app} fetchIcon={() => def} />)
	def.resolve()
	wrapper.update()
	const component = wrapper.getElement()
	expect(component).toMatchSnapshot()
})
```

Reply:
I tested this kind of code when writing this test. I also tested your code but I have not been able to make it work. I guess it is too "quick" maybe ? The only way to make it work was using a `setTimeout`, and God knows I __HATE__ `setTimeout`.

By the way, if you manage to make it works, I am interested by the solution !

Reply:
It would not work with the deferred solution. The reason is that the callbacks for resolution/rejection of a promise are called "asynchronously".

Good article on the subject : https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/